### PR TITLE
Add XSC analytics metrics capabilities

### PR DESCRIPTION
**Pull request by gailazar300**
- [x] The pull request is targeting the `dev` branch.
- [x] The code has been validated to compile successfully by running `go vet ./...`.
- [x] The code has been formatted properly using `go fmt ./...`.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-security/actions/workflows/analysis.yml) passed.
- [x] All [tests](https://github.com/jfrog/jfrog-cli-security/actions/workflows/test.yml) have passed. If this feature is not already covered by the tests, new tests have been added.
- [x] All changes are detailed at the description. if not already covered at [JFrog Documentation](https://github.com/jfrog/documentation), new documentation have been added.

-----

Add analytics metrics capabilities for XSC.
* Adding new XSC server with analytics general events APIs.
* Adding new XSC tests section (for now XSC integration tests were tested and passed with d2c environment which has XSC, skipped in Github-action for now)
* All CLI audit commands will send a request for scanning to XSC instead of Xray from now on.

depends on:
https://github.com/jfrog/jfrog-client-go/pull/928
https://github.com/jfrog/jfrog-cli-core/pull/1165

### CODE REVIEW

File: utils/xscmanager.go
Comment (eranturgeman) [inline utils/xscmanager.go:9]: Please change argument name to serverDetails instead of serviceDetails because server and service mean 2 different things in this context and its best to differentiate between them

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:303]:
```suggestion
func SetAnalyticsMetricsDataForAnalyzerManager(technologies []coreutils.Technology) func() {
	if len(technologies) != 1 {
        // Only report analytics for one technology at a time.
		return
	}
	technology := technologies[0]
	resetAnalyzerManagerPackageManagerVar := SetEnvWithResetCallback(jfPacakgeManagerEnvVariable, technology.String())
	resetAnalyzerManagerLanguageVar := SetEnvWithResetCallback(jfLanguageEnvVariable, technology.Language)
	return func() {
		resetAnalyzerManagerPackageManagerVar()
		resetAnalyzerManagerLanguageVar()
	}
}
```

* create const variables at `analyzermanager.go` and use it
* You can return a func when setting the variables that will reset them
* add general `SetEnvWithResetCallback` to `utils.go` and use it:
```
func SetEnvWithResetCallback(key, value string) func() {
	oldValue, exist := os.LookupEnv(key)
    errMsg := "failed %s %s as environment variable. Cause: %s"

    if err := os.Setenv(key, value); err != nil {
        log.Debug(fmt.Sprintf(errMsg, "setting", key, err.Error()))
        return
    }

	if exist {
		return func() {
            if err := os.Setenv(key, oldValue); err != nil {
                log.Debug(fmt.Sprintf(errMsg, "setting", key, err.Error()))
            }
		}
	}

	return func() {
		if err := os.Unsetenv(key); err != nil {
            log.Debug(fmt.Sprintf(errMsg, "unsetting", key, err.Error()))
        }
	}
}
```

File: commands/audit/audit.go
Comment (attiasas) [inline commands/audit/audit.go:107]: this is a blocking method, consider running this optional operations in diff rotine

File: commands/audit/audit.go
Comment (attiasas) [inline commands/audit/audit.go:184]:
```suggestion
	results.MultiScanId = auditParams.MultiScanId
```
MultiScanId should be stored at auditParams, at this point of adding MSI to every audit scan we can just set the value from the params to the one at the result

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:97]: This method should return the msi and it should be set at the AuditBasicParams (that will set the value of XrayGraphScanParams)
This env var is used to pass the value to the analyzer manager only

File: utils/analyticsmetrics_test.go
Comment (attiasas) [inline utils/analyticsmetrics_test.go:42]:
```suggestion
	cleanReportEnv := clientTests.SetEnvWithCallbackAndAssert(t, coreutils.ReportUsage, "")
    cleanMSIEnv := clientTests.SetEnvWithCallbackAndAssert(t, jfMsiEnvVariable, "")
```

Use clientTests.SetEnvWithCallbackAndAssert

File: utils/analyticsmetrics_test.go
Comment (attiasas) [inline utils/analyticsmetrics_test.go:72]: Define and use Test cases:

testCases := []struct {
    TestName string
    Version string
    expectedResult bool
}{
    ...
}

File: utils/analyticsmetrics_test.go
Comment (attiasas) [inline utils/analyticsmetrics_test.go:88]:
```suggestion
	assert.NotEmpty(t, am.GetMsi())
	assert.Equal(t, params.MultiScanId, am.GetMsi())
```

File: utils/analyticsmetrics_test.go
Comment (attiasas) [inline utils/analyticsmetrics_test.go:98]: can be implemented in test cases, similar code is called.

File: utils/analyticsmetrics_test.go
Comment (attiasas) [inline utils/analyticsmetrics_test.go:157]: what about different in Methods `PUT` and `POST`?

File: go.mod
Comment (attiasas) [inline go.mod:104]: make sure to point to valid remote branch
don't forget to update before merge to point to the dev branch

**Review by attiasas — CHANGES_REQUESTED** [review]
Add MultiScanId to `AuditBasicParams` and use it in as input if MSI was supplied (at Frogbot you can set this param)

Use format tools for the logs.

Describe at the PR description the changes.

More tests are needed

File: go.mod
Comment (gailazar300) [reply go.mod:104]: I have to change the go.mod to run tests with my core and client changes in github-action.

File: utils/analyticsmetrics_test.go
Comment (gailazar300) [reply utils/analyticsmetrics_test.go:88]: what did you mean?

File: utils/analyticsmetrics_test.go
Comment (gailazar300) [reply utils/analyticsmetrics_test.go:72]: this structure is not suitable for this specific test.

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:30]: maybe add it to analyzermanager.go instead? the other env var for it defifned there

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:301]:
```suggestion
	err := os.Setenv(jfPackageManagerEnvVariable, "")
	if err != nil {
		log.Debug(fmt.Sprintf("failed setting %s as environment variable. Cause: %s", jfPackageManagerEnvVariable, err.Error()))
	}
```

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:305]:
```suggestion
	err = os.Setenv(jfLanguageEnvVariable, "")
	if err != nil {
		log.Debug(fmt.Sprintf("failed setting %s as environment variable. Cause: %s", jfLanguageEnvVariable, err.Error()))
	}
```

File: commands/audit/jasrunner.go
Comment (attiasas) [inline commands/audit/jasrunner.go:34]: defer will only set the value after `runJasScannersAndSetResults` is done, after running the Applicability scan...

File: tests/utils/test_utils.go
Comment (attiasas) [inline tests/utils/test_utils.go:25]: check for ValidateXscVersion can be added at the specific test that uses it.
no need to add the check for all tests...

Comment (gailazar300) [reply tests/utils/test_utils.go:25]: there is no separation between xray and XSC tests, there are all security tests, and as far as we get more APIs will approach XSC instead of xray.

File: xsc_test.go
Comment (gailazar300) [reply xsc_test.go:46]: There should be a separation between the regular audit tests and XSC tests, we are testing specific XSC APIs so there should be a separate test section for that and for more XSC tests to come.
This test is specifically restricted by AnalyticsMetricsMinXscVersion, all other audit tests should run without considering XSC.

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:290]:
```suggestion
			if err = resetAnalyzerManageJfMsiVar(); err != nil {
				log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
			}
```
don't ignore err

File: commands/audit/jas/common.go
Comment (attiasas) [inline commands/audit/jas/common.go:311]:
```suggestion
		if err = resetAnalyzerManageJfMsiVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
		}
		if err = resetAnalyzerManagerPackageManagerVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfPackageManagerEnvVariable, err.Error()))
		}
		if err = resetAnalyzerManagerLanguageVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfLanguageEnvVariable, err.Error()))
		}
```
don't ignore err

File: commands/audit/jas/common_test.go
Comment (attiasas) [inline commands/audit/jas/common_test.go:113]: add tests for Zero tech

File: commands/audit/audit.go
Comment (attiasas) [inline commands/audit/audit.go:97]:
```suggestion
	if xscManager := auditCmd.analyticsMetricsService.XscManager(); xscManager != nil {
		version, err := xscManager.GetVersion()
		if err != nil {
			log.Debug(fmt.Sprintf("Can't get XSC version for xray graph scan params. Cause: %s", err.Error()))
		}
		params.XscVersion = version
	}
```

File: tests/utils/test_utils.go
Comment (attiasas) [reply tests/utils/test_utils.go:25]: No need to prepare to that right now, when we will need it, we will change.
This addes more complexity where it is not needed right now

File: tests/utils/test_validation.go
Comment (attiasas) [inline tests/utils/test_validation.go:38]:
```suggestion
	if err = clientUtils.ValidateMinimumVersion(clientUtils.Xsc, xscVersion.GetVersion(), minVersion); err != nil {
		t.Skip(err)
	}
```

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:96]:
```suggestion
	curOs, curArch := "", ""
	if osAndArc, err := coreutils.GetOSAndArc(); err != nil {
		log.Debug(fmt.Errorf("failed to get os and arcitucture for general event request to XSC service, error: %s ", err.Error()))
	} else {
		splitOsAndArch := strings.Split(osAndArc, "-")
		curOs = splitOsAndArch[0]
		curArch = splitOsAndArch[1]
	}
```

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:106]: GetAnalyzerManagerVersion will soon return `[RELEASE]` default value. I think it is neccessery to implement running the command `version` on the analyzerManager binary.

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:109]:
```suggestion
	return &xscservices.XscAnalyticsGeneralEvent{
		XscAnalyticsBasicGeneralEvent: xscservices.XscAnalyticsBasicGeneralEvent{
			EventType:              eventType,
			EventStatus:            xscservices.Started,
			Product:                product,
			JfrogUser:              ams.xscManager.Config().GetServiceDetails().GetUser(),
			OsPlatform:             curOs,
			OsArchitecture:         curArch,
			AnalyzerManagerVersion: GetAnalyzerManagerVersion(),
		},
	}
```

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:114]:
```suggestion
		log.Debug("Analytics metrics are disabled, skipping sending event request to XSC")
```

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:132]:
```suggestion
	if !ams.ShouldReportEvents() {
		return
	}
```
No need to log at the end, we already logged that it is disabled

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:134]:
```suggestion
		log.Debug("MultiScanId is empty, skipping update general event...")
```

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:140]:
```suggestion
	if err := ams.xscManager.UpdateAnalyticsGeneralEvent(*event); err != nil {
		log.Debug(fmt.Sprintf("failed updading general event request in XSC service for multi_scan_id %s, error: %s \"", ams.GetMsi(), err.Error()))
	} else {
        log.Debug(fmt.Sprintf("General event updated\n%s", *event)
    }
```
We can show the summary information on debug level.

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:152]:
```suggestion
	return ams.xscManager.GetAnalyticsGeneralEvent(msi)
```
log can be added where this func is used, we return it...

File: utils/analyticsmetrics.go
Comment (attiasas) [inline utils/analyticsmetrics.go:174]: I would consider adding this logic as methods in the `Results` struct

File: utils/mocks.go
Comment (attiasas) [inline utils/mocks.go:1]: move this file to `tests/utils` because it contains `t *testing.T` please rename this file to `test_mocks.go`

File: go.mod
Comment (attiasas) [inline go.mod:103]:
```suggestion
replace github.com/jfrog/jfrog-cli-core/v2 => github.com/jfrog/jfrog-cli-core/v2 dev

replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
```
Update to dev branch, those changes already merged

File: xsc_test.go
Comment (attiasas) [inline xsc_test.go:47]: No need for specific file for analytics integration and add another scan for that. it can be added to an existing test at `audit_test.go` for example `TestXrayAuditJasSimpleJson` since it will also have JAS issues. the func `validateAnalyticsBasicEvent` can be moved to `tests/utils/test_validation.go`

**Review by attiasas — COMMENTED** [review]
Please make sure it is working on platform that don't have XSC

File: tests/utils/test_validation.go
Comment (gailazar300) [reply tests/utils/test_validation.go:38]: The way I implemented it is totally fine as well.

File: commands/audit/audit.go
Comment (gailazar300) [reply commands/audit/audit.go:97]: The way I implemented it is totally fine as well.

File: utils/analyticsmetrics.go
Comment (gailazar300) [reply utils/analyticsmetrics.go:96]: its the same.

File: utils/analyticsmetrics.go
Comment (gailazar300) [reply utils/analyticsmetrics.go:109]: The way I implemented it is totally fine as well.

File: utils/analyticsmetrics.go
Comment (gailazar300) [reply utils/analyticsmetrics.go:106]: The person who will implement it in the future will do so when it is relevant.

File: utils/analyticsmetrics.go
Comment (gailazar300) [reply utils/analyticsmetrics.go:132]: I think that its ok to add in debug mode also if we are skipping the updating.

File: utils/analyticsmetrics.go
Comment (gailazar300) [reply utils/analyticsmetrics.go:152]: this function is for testing and getting the analytics from xsc - I think that its fine like this.

File: utils/mocks.go
Comment (gailazar300) [reply utils/mocks.go:1]: this caused circle import.
the unit tests are using this mock, so this struct can be in theire directory.

File: go.mod
Comment (gailazar300) [reply go.mod:103]: again- of course, I wont merge the go.mod like this!
and will run tests with the dev branches, no need to write it every time.

File: xsc_test.go
Comment (gailazar300) [reply xsc_test.go:47]: again as I mentioned before - I think that it's necessary to have a separate section for XSC testing and to have a test that tests analytics and nothing else.
I don't think that adding this in the audit section in one random test and validating analytics after other things will be right and clear.

Comment (gailazar300) [reply xsc_test.go:47]: and I will add here also about "InitSecurityTest" - we are going to work mostly with XSC instead of xray for now on - so now when the infrastructure is still small it's time to add testing xsc support.

**Review by attiasas — APPROVED** [review]

**Pull request merged**
